@xen-orchestra/audit-core/src/index.js

```diff
@@ -0,0 +1,121 @@
+// see https://github.com/babel/babel/issues/8450
+import 'core-js/features/symbol/async-iterator'
+
+import assert from 'assert'
+import hash from 'object-hash'
+import { invert } from 'lodash'
+
+export class Storage {
+  constructor() {
+    this.locked = undefined
+  }
+
+  async lock() {
+    if (this.locked !== undefined) {
+      await this.locked
+    }
+
+    let res
+    this.locked = new Promise(resolve => {
+      res = resolve
+    })
+    return () => {
+      this.locked = undefined
+      return res()
+    }
+  }
+}
+
+// Format: $<algorithm>$<salt>$<encrypted>
+//
+// http://man7.org/linux/man-pages/man3/crypt.3.html#NOTES
+const ALGORITHM_TO_ID = {
+  sha256: '5',
+}
+
+const ID_TO_ALGORITHM = invert(ALGORITHM_TO_ID)
+
+export const NULL_ID = 'nullId'
+
+const HASH_ALGORITHM = 'sha256'
+const createHash = (data, algorithm = HASH_ALGORITHM) =>
+  `$${ALGORITHM_TO_ID[algorithm]}$$${hash(data, {
+    algorithm,
+    excludeKeys: key => key === 'id',
+  })}`
+
+export class AuditCore {
+  constructor(storage) {
+    assert(storage !== undefined && storage.lock !== undefined)
+    this._storage = storage
+  }
+
+  async add(subject, event, data) {
+    const storage = this._storage
+    const unlock = await storage.lock()
+    const record = {
+      data,
+      event,
+      previousId: (await storage.getLastId()) ?? NULL_ID,
+      subject,
+      time: Date.now(),
+    }
+    record.id = createHash(record)
+    await storage.put(record)
+    await storage.setLastId(record.id)
+    unlock()
```

Comment:
Should it be `async` or not?

Reply:
We don't need to wait for the release of the lock, the `add` method should finish when the record is properly stored.

Reply:
Can `unlock` throw or reject?

Reply:
No, it'll all time resolves.

Reply:
*resolves*? Does that mean it's an async function?

Keep in mind, we are not talking about your implementation, but about the general specification for this function.

Reply:
It's not an sync function it just resolve a promise.